README
Please see the ./output/index.html file in your Web browser.


HISTORY
--
Note: A HTML version of this file is available in the doc/ directory.
++


= 1.0.0 (2006-12-11)

This is the very first release of Rassmalog; happy birthday!


config/format.rb
# This file defines the String#to_html method which is invoked to transform
# the content of an entry (the value of the string) into HTML.
#
# It features, in addition to the Textile formatting system (RedCloth) and
# syntax coloring (CodeRay), smart source code sizing (block vs. inline
# display) and table of contents generation.

=begin
  Copyright 2006 Suraj N. Kurapati

  This file is part of Rassmalog.

  Rassmalog is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  Rassmalog is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Rassmalog; if not, write to the Free Software Foundation,
  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
=end

require 'cgi'

begin
  require 'rubygems'
rescue LoadError
end

require 'coderay'
require 'redcloth'

class String
  # The content of these HTML tags will be preserved verbatim when they are processed by Textile.
  PRESERVED_TAGS = [:code, :tt]

  # Transforms this string into HTML.
  def to_html
    text = dup

    # prevent the content of these tags from being transformed by Textile
    # for example, Textile transforms quotation marks in code into curly ones (&#8192;) -- this ruins any source code in the content of the tags!
      PRESERVED_TAGS.each do |tag|
        text.gsub! %r{<#{tag}(.*?)>(.*?)</#{tag}>}m, %{<pre tag=#{tag.inspect}\\1>\\2</pre>}
      end

    html = text.redcloth

    # restore the original tags for the preserved tags
      # unescape content of <pre> tags because they may contain nested preserved tags (redcloth escapes the content of <pre> tags)
        html.gsub! %r{(<pre>)(.*?)(</pre>)}m do
          $1 + CGI.unescapeHTML($2) + $3
        end

      PRESERVED_TAGS.each do |tag|
        html.gsub! %r{<pre tag=#{tag.inspect}(.*?)>(.*?)</pre>}m, %{<#{tag}\\1>\\2</#{tag}>}
      end

      # assume that indented text in Textile is NOT source code
        html.gsub! %r{(<pre>)\s*<code>(.*?)\s*</code>\s*(</pre>)}m, '\1\2\3'

      # escape content of <pre> tags, because we un-escaped it above
        html.gsub! %r{(<pre>)(.*?)(</pre>)}m do
          $1 + CGI.escapeHTML($2) + $3
        end

    html.coderay
  end

  # Returns the result of running this string through RedCloth.
  def redcloth
    RedCloth.new(self).to_html
  end

  # Adds syntax coloring to <code> elements in the given text. If
  # the <code> tag has an attribute lang="...", then that is considered the
  # programming language for which appropriate syntax coloring should be
  # applied. Otherwise, the programming language is assumed to be ruby.
  def coderay
    gsub %r{<(code)(.*?)>(.*?)</\1>}m do
      code = CGI.unescapeHTML $3
      atts = $2

      lang =
        if $2 =~ /lang=('|")(.*?)\1/i
          $2
        else
          :ruby
        end

      tag =
        if code =~ /\n/
          :pre
        else
          :code
        end

      html = CodeRay.scan(code, lang).html(:css => :style)

      %{<#{tag} class="code"#{atts}>#{html}</#{tag}>}
    end
  end


  Heading = Struct.new :anchor, :title, :depth
  @@anchorNum = 0

  # Builds a table of contents from text formatted in Textile.
  # Returns an array containing [toc, text] where:
  #
  # toc::   the generated table of contents (formatted in Textile)
  #
  # text::  a modified version of this string which contains anchors for the
  #         hyperlinks in the table of contents (so that the TOC can link to
  #         the content in this string)
  #
  def table_of_contents
    headings = []

    # parse document structure and insert anchors (so that the table of contents can link directly to these headings) where necessary
      text = gsub %r{^(\s*h(\d))(.*)$} do
        head, depth, rest = $1, $2, $3

        # parse title and class attributes
          rest =~ /^([\{\(\[].*?[\]\)\}])?\.(.*)$/
          atts, title = $1, $2.strip

        # parse and insert anchor if necessary
          if atts =~ /#(.*?)\)/
            anchor = $1
          else
            anchor = "anchor#{@@anchorNum += 1}"
            rest.insert 0, "(##{anchor})"
          end

        headings << Heading.new(anchor, title.strip, depth.to_i)

        head + rest
      end

    # generate table of contents
      toc = headings.map do |h|
        %{#{'*' * h.depth} "#{h.title}":##{h.anchor}}
      end.join("\n").redcloth

    [toc, text]
  end
end


lib/blog.rb
# This file contains the logic for generating the blog. It is designed to be
# embedded inside the main Rakefile in the parent directory because all
# paths in this file are relative to the parent directory.
=begin
  Copyright 2006 Suraj N. Kurapati

  This file is part of Rassmalog.

  Rassmalog is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  Rassmalog is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Rassmalog; if not, write to the Free Software Foundation,
  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
=end

require 'rake/clean'
require 'config/format'
require 'yaml'
require 'ostruct'
require 'date'

require 'erb'
include ERB::Util

ERB_CONTEXT = self


class DateTime
  # Returns the RFC-822 representation, which is required by RSS, of this object.
  def rfc822
    strftime "%a, %d %b %Y %H:%M:%S %Z"
  end
end

class String
  # Transforms this string into a vaild file name.
  def to_file_name
    # file names cannot have slash
    gsub '/', '|'
  end
end


## data structures for organizing entries

Page = Struct.new :name, :url

# Represents a generated HTML page, which is associated with a list of entries. This association is maintained by the Chapter class.
class Page
  alias old_url url

  # Returns a relative URL to this page.
  def url
    (old_url || name).to_s.to_file_name << '.html'
  end

  # Returns a relative hyperlink to this page.
  def to_link aName = name
    %{<a href="#{u url}">#{aName}</a>}
  end

  # Compares this page to the given page.
  def <=> aOther
    url <=> aOther.url
  end

  alias to_s name
end

# A mapping from a Page to its entries.
class Chapter < Hash
  attr_reader :name

  # aName:: name of this chapter
  # the rest:: arguments to Hash.new
  def initialize aName, *aArgs, &aBlock
    @name = aName
    super *aArgs, &aBlock
  end

  alias pages keys

  # Renders, within the context of the given blog, the given page into HTML.
  def render aPage
    entries = self[aPage]
    title = "#{name} &mdash; #{aPage.name}"

    ERB_CONTEXT.instance_eval do
      @page_title = title
      heading = "<h2>#{@page_title}</h2>\n\n"

      @page_content = entries.inject heading do |memo, entry|
        memo << entry.to_html(true)
      end

      HTML_TEMPLATE.result(binding)
    end
  end
end

module Entry
  # Returns the name of the generated HTML file.
  def url
    stamp = date.strftime "%F"
    "#{stamp}-#{name}.html".to_file_name
  end

  # Returns a hyperlink to the generated HTML file.
  def to_link aName = name
    %{<a href="#{u url}">#{aName}</a>}
  end

  # Transforms this entry into HTML (for the generated HTML file).
  def to_html aSummarize = false
    old = text

    # summarize the entry body
      if aSummarize and text =~ /^.*?(\r?\n){2,}/m
        link = to_link "Read the rest of this entry &rarr;"
        self.text = $& << link
      end

    # transform the entry into HTML
      @entry = self
      html = ENTRY_TEMPLATE.result binding

    self.text = old
    html
  end

  # Renders, within the context of the given blog, a HTML page for this entry.
  def render
    t, c = name, to_html
    ERB_CONTEXT.instance_eval do
      @page_title = t
      @page_content = c
      HTML_TEMPLATE.result binding
    end
  end

  # Compares this entry to the given entry.
  def <=> aOther
    date <=> aOther.date
  end
end


# Notify the user about some action being performed.
def notify *args
  printf "%8s  %s\n", *args
end

# Loads the given YAML file into an OpenStruct.
def load_yaml_file aFile
  OpenStruct.new(YAML.load_file(aFile))
end

# Writes the given content to the given file.
def write_file aPath, aContent
  File.open aPath, 'w' do |f|
    # lstrip because XML declaration must be at start of file
    f << aContent.lstrip
  end
end


## input processing stage

# load blog configuration
  @blog = load_yaml_file('config/blog.yaml')

  FileList['config/*.erb'].each do |f|
    name = File.basename(f, File.extname(f))
    var = "#{name.upcase}_TEMPLATE"

    Kernel.const_set var.to_sym, ERB.new(File.read(f))
  end

# load blog entries
  @entries = FileList['entries/**/*.{yml,yaml}'].map do |src|
    entry = load_yaml_file(src)
    entry.src_file = src
    entry.date = DateTime.parse(entry.date.to_s)
    entry.tags = entry.tags.to_a rescue [entry.tags]
    entry.extend Entry

    entry
  end.sort.reverse!

# organize blog entries into chapters
  @tags = Chapter.new("Tags") {|h,k| h[k] = []}
  @archives = Chapter.new("Archives") {|h,k| h[k] = []}

  # this stuff is done *after* the entries have been sorted, so that stuff in the archives appears in the correct chronological order
  @entries.each do |entry|
    # determine which tags this entry belongs to
      entry.tags.map! do |tag|
        page = Page.new tag
        @tags[page] << entry

        page
      end

    # determine which archive this entry belongs to
      arch = Page.new entry.date.strftime("%B %Y"), entry.date.strftime("%Y-%m")

      @archives[arch] << entry
  end

  @chapters = [@tags, @archives]


## output generation stage

desc "Generate the blog."
task :blog

COMMMON_DEPS = FileList['config/*', 'output']

# generate output directory
  directory 'output'
  CLOBBER.include 'output'

  # copy everything from input/ into output/
    FileList['input/**/*'].each do |src|
      dst = "output/#{File.basename src}"

      file dst => [src, 'output'] do
        cp_r src, dst, :preserve => true
      end

      task :blog => dst
      CLEAN.include dst
    end

# generate pages for entries
  @entries.each do |entry|
    dst = entry.dst_file = File.join('output', entry.url)

    file dst => [entry.src_file, *COMMMON_DEPS] do
      write_file dst, entry.render
      notify :entry, dst
    end

    task :blog => dst
    CLOBBER.include dst
  end

# generate archive pages for entries
  index = Chapter.new 'Index'
  index[Page.new('newest entries', 'index')] = @entries[0, @blog.index]

  (@chapters + [index]).each do |chapter|
    chapter.each_pair do |page, entries|
      dst = File.join('output', page.url)
      deps = entries.map {|e| e.dst_file} << 'output'

      file dst => deps do
        write_file dst, chapter.render(page)
        notify chapter.name, dst
      end

      task :blog => dst
      CLOBBER.include dst
    end
  end

# generate RSS feed
  file 'output/rss.xml' => COMMMON_DEPS do |t|
    write_file t.name, RSS_TEMPLATE.result(binding)
    notify :rss, t.name
  end

  task :blog => 'output/rss.xml'
  CLOBBER.include 'output/rss.xml'
